File: Courses/Computational_Techniques/SVD.m
m = 250;
n = 150;
r = 100;
l = 5;

A = rand(m, n);

Omega = rand(n, r + l);
[Q R] = qr(A * Omega);
[U S V] = svd(Q' * A);

U = Q * U;
U = U(:,1:r);
S = S(1:r,1:r);
V = V(:,1:r);

Arand = U * S * V';

[Ur Sr Vr] = svd(A);

Ur = U(:,1:r);
Sr = S(1:r,1:r);
Vr = V(:,1:r);

Ar = Ur * Sr * Vr';

norm(A - Arand) / norm(A - Ar)

File: Courses/Computational_Techniques/SVD.py
import numpy as np

def rand_SVD(A, r, l):
	m, n = A.shape
	Omega = np.random.normal(size = (n, r + l))
	Q, R = np.linalg.qr(np.matmul(A, Omega))
	U, Sigma, V = np.linalg.svd(np.matmul(Q.T, A), full_matrices = False)
	return np.matmul(Q, U)[:, :r], Sigma[:r], V[:r]

r = 140
A = np.random.normal(size = (250, 150))

u, s, v = rand_SVD(A, r, 5)
Arand = np.matmul((u * s), v)


ur, sr, vr = np.linalg.svd(A, full_matrices = False)
Ar = np.matmul((ur[:, :r] * sr[:r]), vr[:r])

print sr[r]
print np.linalg.norm(A - Ar, ord = 2)
print np.linalg.norm(A - Arand, ord = 2)

print np.linalg.norm(A - Arand, ord = 2) / np.linalg.norm(A - Ar, ord = 2)


